In [47]:
import requests
import urllib.request
from bs4 import BeautifulSoup
import pandas as pd
import re
import numpy as np

We started by analyzing the html page and locating the data that we need i.e the ranking, country and region, total and international academic stuff, as well as total and international number of students. As mentioned in the homework description the information about the number of staff and students is located in a page other than the one that provides the rank ordering. A redirection to that page is provided by a link on the name of the university and is provided by the root/default/home url suffixed with "/university/some-format-of-the-university-name". Our next task is finding a way to collect the data. 

In order to discover the way the webpage fetches the data we used postman with its interceptor extension, which enabeled us to monitor the requests generated upon every navigation to the given web page. In particular we learned that the data in our interest is returned from an API with a get request to the *data_url* (used in the cell below) in json format. After inspecting the response json object, it has been noticed that all the data displayed on the rankings page originates precisely from this source, and is available in a convinient format. Even more a link to the individual webpage for every university is available, which gives us a way to collect the rest of the data we need. That data though has to be collected by crawling the html response.

In [3]:
# the home URL used for generating the URLs to every individual university site
root_url = 'http://www.topuniversities.com'
# the URL to the API that returns the university data used to generate the table/list of universities (mentioned above)
data_url = 'https://www.topuniversities.com/sites/default/files/qs-rankings-data/357051.txt?_=1508229310180'
data_request = requests.get(data_url)

In [4]:
# parsing the data and extracting the list of universities from the json
data_json = data_request.json()
ordered_list_of_universities = data_json['data']

Now as we have all that we have all of the data at reach, we just parse it in the following manner. 

First we extract the data available directly in the dictionay such as title, country, region and rank. The rank in particular with using a regex is trimmed from all of the possible nondigit signes present in it, in order to be succesfully parsed as an int as is should be (this is a more general solution that will work in wider spectrum of problems, while our motivation is eliminating the '=' symbols present to indicate a situation where more than one university is given the same rank i.e its a draw).

Next, we use the URL in the dictionary that leads to the individual website for the university, which in turn after crawling it's HTML provides us with the rest of the information we need. After inspection, the nodes in the DOM that contain the numbers of interest are located, and using given CSS selectors manipulated and extracted.

In [6]:
i = 0
university_parsed_data = []
for uni_data in ordered_list_of_universities:
    # This try is present in order to catch any potentional error while parsing and print it nicely
    try: 
        uni_name = uni_data['title']
        uni_country = uni_data['country']
        uni_region = uni_data['region']
        uni_rank = int(re.sub(r'^[^\d]+|[^\d]+$', '', uni_data['rank_display']))
        uni_url = uni_data['url']

        uni_url_request = requests.get("%s%s" % (root_url, uni_url))
        soup = BeautifulSoup(uni_url_request.text, 'html.parser')

        # Usually two objects are selected, since the website has mobile view and desktop view of the table
        # For some universities this information is not available, hence we save the value as -1 to note that
        # In the selection it is noted that all of the numbers of interest are contained in a div with class number, that
        # in turn is always contained in another parent div whose class somehow informs us about its semantical meaning
        # We use the classes and node types in CSS selectors to get the information of interest.
        # The information is trimmed of white spaces and transformed in a suitable format
        total_afs = soup.select('div.total.faculty div.number')
        if len(total_afs) == 0:
            total_afs = -1
        else:
            total_afs = total_afs[0].text.strip().replace(',', '') # total academic faculty stuff
            
        international_afs = soup.select('div.inter.faculty div.number')
        if len(international_afs) == 0:
            international_afs = -1
        else:
            international_afs = international_afs[0].text.strip().replace(',', '') # international academic faculty staff number
        
        total_students = soup.select('div.total.student div.number')
        if len(total_students) == 0:
            total_students = -1
        else:
            total_students = total_students[0].text.strip().replace(',', '') # total number of students

        inter_students = soup.select('div.total.inter div.number')
        if len(inter_students) == 0:
            inter_students = -1
        else:
            inter_students = inter_students[0].text.strip().replace(',', '') # international students number
        
        # Combine all of the collected data in a concise dictionary, and append it to a list later used to create a dataframe
        university_parsed_data.append({'Rank': uni_rank, 'Name': uni_name, 'Country': uni_country,  'Region': uni_region,
                        'Total_AFS': int(total_afs), 'International_AFS': int(international_afs), 
                        'Total_students': total_students, 'International_students': inter_students})
    except ValueError as e:
        print(e)
        print("ERROR")
        break
    i += 1
    if i == 200:
        print("SUCCESS")
        break

SUCCESS


In [8]:
# NaN values are set to -1
# There is no university ranked 198, in particular the ranking goes from 3 universities ranked 195 and then continues to 199 
data = pd.DataFrame(university_parsed_data)[['Rank', 'Name', 'Country', 'Region', 
                        'Total_AFS', 'International_AFS', 
                        'Total_students', 'International_students']]
data['Total_students'] = data.Total_students.astype(np.int64)
data['International_students'] = data.International_students.astype(np.int64)
data['Total_AFS'] = data.Total_AFS.astype(np.int64)
data['International_AFS'] = data.International_AFS.astype(np.int64)
data.head()

Rank                                          Name         Country  \
0     1   Massachusetts Institute of Technology (MIT)   United States   
1     2                           Stanford University   United States   
2     3                            Harvard University   United States   
3     4  California Institute of Technology (Caltech)   United States   
4     5                       University of Cambridge  United Kingdom   

          Region  Total_AFS  International_AFS  Total_students  \
0  North America       2982               1679           11067   
1  North America       4285               2042           15878   
2  North America       4350               1311           22429   
3  North America        953                350            2255   
4         Europe       5490               2278           18770   

   International_students  
0                    3717  
1                    3611  
2                    5266  
3                     647  
4                    6699

Since it is noticed that some of the part is repetative, a function is defined in order to provent code repetition.

This function gets a *DataFrame* and two *column names*, removes all the rows that contain none in either of them, and than calculates the ratio between the values in these columns for each row of the dataframe, while firstly grouping the data of the dataframe by a given column if a column of the dataframe is passed to the call to group by. The returning DataFrame is sorted by the value of the calculated ratio, which is joined to every row in a column passed as the parameter *new_column_name*. 

It is assumed that the passed *DataFrame* has columns anmed *col1*, *col2* and *group_by* (if used), as well as that the values for *col1*, *col2* are numberic.

In [65]:
def find_ratio(df, col1, col2, new_column_name, group_by=None, ascending=False):
    no_nan_in_columns_df = df[(df[col1] != -1) & (df[col2] != -1)]
    if group_by != None:
        no_nan_in_columns_df = no_nan_in_columns_df.groupby(group_by)[[col1, col2]].sum()
    sorted_ratio = (no_nan_in_columns_df[col1] / no_nan_in_columns_df[col2]).sort_values(ascending=ascending)
    return no_nan_in_columns_df.loc[sorted_ratio.index].join(pd.DataFrame(sorted_ratio, columns=[new_column_name]))

In [78]:
# Calculate the ratio between the number students and staff members, and sort the dataframe according to that value
# It is assumed that the smaller the ratio the better, since there are less students per teacher
find_ratio(data, 'Total_students', 'Total_AFS', 'Ratio', ascending=True).head(10)

Rank                                          Name         Country  \
3       4  California Institute of Technology (Caltech)   United States   
15     16                               Yale University   United States   
5       6                          University of Oxford  United Kingdom   
4       5                       University of Cambridge  United Kingdom   
16     17                      Johns Hopkins University   United States   
1       2                           Stanford University   United States   
0       1   Massachusetts Institute of Technology (MIT)   United States   
185   186                       University of Rochester   United States   
18     19                    University of Pennsylvania   United States   
17     18                           Columbia University   United States   

            Region  Total_AFS  International_AFS  Total_students  \
3    North America        953                350            2255   
15   North America       4940               1708           12402   
5           Europe       6750               2964           19720   
4           Europe       5490               2278           18770   
16   North America       4462               1061           16146   
1    North America       4285               2042           15878   
0    North America       2982               1679           11067   
185  North America       2569                488            9636   
18   North America       5499               1383           20639   
17   North America       6189                913           25045   

     International_students     Ratio  
3                       647  2.366212  
15                     2469  2.510526  
5                      7353  2.921481  
4                      6699  3.418944  
16                     4105  3.618557  
1                      3611  3.705484  
0                      3717  3.711268  
185                    2805  3.750876  
18                     4250  3.753228  
17                     8105  4.046696

In [11]:
# Calculate the ratio between the number of international students and  total number of students, sort the dataframe according to that value
# It is assumed that we are interested in the universities that have higher number of international students
find_ratio(data, 'International_students', 'Total_students', 'Ratio').head(10)

Rank                                               Name         Country  \
34     35  London School of Economics and Political Scien...  United Kingdom   
11     12    Ecole Polytechnique Fédérale de Lausanne (EPFL)     Switzerland   
7       8                            Imperial College London  United Kingdom   
198   200                              Maastricht University     Netherlands   
47     47                         Carnegie Mellon University   United States   
6       7                    UCL (University College London)  United Kingdom   
91     92                           University of St Andrews  United Kingdom   
41     41                        The University of Melbourne       Australia   
126   127                    Queen Mary University of London  United Kingdom   
25     26                        The University of Hong Kong       Hong Kong   

            Region  Total_AFS  International_AFS  Total_students  \
34          Europe       1088                687            9760   
11          Europe       1695               1300           10343   
7           Europe       3930               2071           16090   
198         Europe       1277                502           16385   
47   North America       1342                425           13356   
6           Europe       6345               2554           31080   
91          Europe       1140                485            8800   
41         Oceania       3311               1477           42182   
126         Europe       1885                801           16135   
25            Asia       3012               2085           20214   

     International_students     Ratio  
34                     6748  0.691393  
11                     5896  0.570047  
7                      8746  0.543567  
198                    8234  0.502533  
47                     6385  0.478062  
6                     14854  0.477928  
91                     4030  0.457955  
41                    18030  0.427434  
126                    6806  0.421816  
25                     8230  0.407144

In [84]:
# Group the data according to the value of the country attribute
# Calculate the ratio between the number of students and staff members, sort the dataframe according to that value
find_ratio(data, 'Total_students', 'Total_AFS', 'Ratio', 'Country', ascending=True).head()

Total_students  Total_AFS     Ratio
Country                                          
Russia                 30233       6709  4.506335
Denmark                67223      11916  5.641407
Saudi Arabia            6040       1062  5.687382
Singapore              58466       9444  6.190809
Malaysia               17902       2755  6.498004

In [13]:
# Group the data according to the value of the country attribute
# Calculate the ratio between the number of international students and total number of students, sort the dataframe according to that value
find_ratio(data, 'International_students', 'Total_students', 'Ratio', 'Country').head()

International_students  Total_students     Ratio
Country                                                         
Australia                       106359          301994  0.352189
United Kingdom                  199426          583621  0.341705
Hong Kong                        24499           78838  0.310751
Austria                          19667           63446  0.309980
Switzerland                      32995          109112  0.302396

In [114]:
# Group the data according to the value of the region attribute
# Calculate the ratio between the number of students and staff members, sort the dataframe according to that value
find_ratio(data, 'Total_students', 'Total_AFS', 'Ratio', 'Region', ascending=True)

Total_students  Total_AFS      Ratio
Region                                             
Asia                   807003     106734   7.560880
North America         1546353     182123   8.490707
Europe                1957251     218358   8.963496
Latin America          435750      45382   9.601825
Africa                  19593       1733  11.305828
Oceania                350167      25347  13.814929

In [91]:
# Group the data according to the value of the region attribute
# Calculate the ratio between the number of international students and total number of students, sort the dataframe according to that value
find_ratio(data, 'International_students', 'Total_students', 'Ratio', 'Region')

International_students  Total_students     Ratio
Region                                                         
Oceania                        118798          350167  0.339261
Europe                         449364         1957251  0.229589
North America                  292116         1546353  0.188906
Africa                           3325           19593  0.169703
Asia                           110100          807003  0.136431
Latin America                   36871          435750  0.084615

### 2. Obtain the 200 top-ranking universities in www.timeshighereducation.com (ranking 2018). Repeat the analysis of the previous point and discuss briefly what you observed.

By analysing the traffic in the browser's inspection tools, it can easily be noticed that when we try to open https://www.timeshighereducation.com, the browser sends a `GET` request for a `JSON` file that contains the entire ranking of universities that is then rendered in the `HTML` of the website. The most practical way of obtaining that information is to use Python to send a request to fetch the `JSON` formatted file from the `URL` we see in the command line tool of the browser.

In [18]:
base_url = 'https://www.timeshighereducation.com'
ranking_data_json_path = '/sites/default/files/the_data_rankings/world_university_rankings_2018_limit0_369a9045a203e176392b9fb8f8c1cb2a.json'

times_ranking_json = requests.get(base_url + ranking_data_json_path).json()

In [75]:
times_ranking = times_ranking_json['data']

times_ranking_array = []
country_region_mapping = {}
# The case of South Korea needs to be treated manually since the official name of the country is
# Republic of Korea, and it's not recognised by the API service
country_region_mapping['South Korea'] = 'Asia'
for uni_object in times_ranking:
    # Just in case a parsing error occurs
    try:
        # Universities ranked with a range, ex. a–b, can be considered equally ranked as a
        # since that is the most information we can get, but it simplifies the processing
        uni_rank = int(re.sub('=', '', uni_object['rank'].split('–')[0]))
        
        # We need only the top 200 universities
        if uni_rank > 200:
            break
            
        uni_name = uni_object['name']
        uni_country = uni_object['location']
        uni_url = uni_object['url']
        
        # The absolute number of international students is not given, but it can easily be computed
        # based on the percentage of international students and the total number of students 
        uni_total_students = int(re.sub(',', '', uni_object['stats_number_students'])) # Remove the thousands separator
        uni_percentage_international = float(re.sub('%', '', uni_object['stats_pc_intl_students'])) / 100.0
        uni_international_students = int(round(uni_total_students * uni_percentage_international))
        
        # The absolute number of faculty is not given, but there is a student-teacher ratio in the data
        # which is basically the number of students divided by the number of staff, so a formula can be derived
        # The numbers differ greatly from what has been observed from the first data source
        uni_student_faculty_ratio = float(uni_object['stats_student_staff_ratio'])
        uni_total_staff = int(round(uni_total_students / uni_student_faculty_ratio))
        
        # The region of the university is not given in the data, but there is a very simple public API
        # that can be used to fetch the region information based on the country name, called Rest Countries
        rest_country_url = "https://restcountries.eu/rest/v2/name/%s?fields=region" % urllib.request.pathname2url(uni_country)
        # Cache the results instead of sending GET requests all the time
        if uni_country in country_region_mapping:
            uni_region = country_region_mapping[uni_country]
        else:
            uni_region_raw = requests.get(rest_country_url).json()
            uni_region = None
            # If it's not a list then it's an error and the region could not be found
            if type(uni_region_raw) is list:
                uni_region = uni_region_raw[0]['region']
            # Cache the new result
            country_region_mapping[uni_country] = uni_region
        
        times_ranking_array.append({
            'Rank': uni_rank,
            'Name': uni_name,
            'Country': uni_country,
            'Region': uni_region,
            'International_faculty': None,
            'Total_faculty': uni_total_staff,
            'International_students': uni_international_students,
            'Total_students': uni_total_students
        })
        
    except ValueError as e:
        print(e)
        print("ERROR")
        break

In [76]:
times_ranking_df = pd.DataFrame(times_ranking_array)[['Rank', 'Name', 'Country', 'Region',
                                                      'Total_faculty', 'International_faculty',
                                                      'Total_students', 'International_students']]
times_ranking_df.head()

Rank                                   Name         Country    Region  \
0     1                   University of Oxford  United Kingdom    Europe   
1     2                University of Cambridge  United Kingdom    Europe   
2     3     California Institute of Technology   United States  Americas   
3     3                    Stanford University   United States  Americas   
4     5  Massachusetts Institute of Technology   United States  Americas   

   Total_faculty International_faculty  Total_students  International_students  
0           1822                  None           20409                    7755  
1           1687                  None           18389                    6436  
2            340                  None            2209                     596  
3           2113                  None           15845                    3486  
4           1285                  None           11177                    3800

The data that Times Higher Education ranking website provides is not as informative as the one from Top Universities, especially in the categories that are requested in the assignment. It is worse in terms that it does not contain raw data and raw numbers, but aggregated ones or data that has already undergone some processing and calculations.

For example: 
* the number of international students is given as a percentage of the total number, so the absolute value needs to be computed. 
* there is no information about international faculty (or staff) at the universities, neither as raw data nor percentage, i.e. there is no way to deduce or calculate the number
* there is no region information for a particular university or country, therefore another source of information was needed, that might not provide the exact same region labels as Top Universities, but in the end the numbers in the groupings are easily comparable
* it was necessary to compute the total faculty based on the number of students and the student-teacher ratio using the following relationship between the variables:
$$student\_teacher\_ratio = \frac{\#students}{\#staff}$$

** Best universities in terms of: ratio between faculty members and students **

In [77]:
# Calculate the student teacher ratio
# It is assumed that the smaller the ratio the better, since there are less students per teacher
find_ratio(times_ranking_df, 'Total_students', 'Total_faculty', 'Student_teacher_ratio', ascending=True).head(10)

Rank                              Name        Country    Region  \
105   105             Vanderbilt University  United States  Americas   
109   109          University of Copenhagen        Denmark    Europe   
11     12                   Yale University  United States  Americas   
153   153           University of Rochester  United States  Americas   
12     13          Johns Hopkins University  United States  Americas   
97     98                  Emory University  United States  Americas   
16     17                   Duke University  United States  Americas   
114   115               École Polytechnique         France    Europe   
183   184  Scuola Normale Superiore di Pisa          Italy    Europe   
101   100          University of Pittsburgh  United States  Americas   

     Total_faculty International_faculty  Total_students  \
105           3640                  None           12011   
109           7413                  None           30395   
11            2827                  None           12155   
153           2241                  None            9636   
12            3604                  None           15498   
97            2875                  None           12649   
16            3390                  None           15256   
114            549                  None            2799   
183             99                  None             515   
101           4469                  None           26367   

     International_students  Student_teacher_ratio  
105                    1561               3.299725  
109                    4255               4.100229  
11                     2553               4.299611  
153                    2794               4.299866  
12                     3720               4.300222  
97                     2403               4.399652  
16                     3356               4.500295  
114                    1008               5.098361  
183                      36               5.202020  
101                    2637               5.899978

Even though the ratio numbers are very off, Yale University, Johns Hopkins University and University of Rochester, all in the US, are among the best 10 according to student-teacher-ratio.

** Best universities in terms of: ratio of international students **

In [79]:
# Calculate the ratio between the number of international students and total number of students
# and sort the dataframe according to that value
# It is assumed that we are interested in the universities that have high percentage of international students
find_ratio(times_ranking_df, 'International_students', 'Total_students', 'Internationality_ratio').head(10)

Rank                                              Name         Country  \
24     25  London School of Economics and Political Science  United Kingdom   
178   179                          University of Luxembourg      Luxembourg   
7       8                           Imperial College London  United Kingdom   
37     38          École Polytechnique Fédérale de Lausanne     Switzerland   
102   103                             Maastricht University     Netherlands   
15     16                         University College London  United Kingdom   
143   143                          University of St Andrews  United Kingdom   
23     24                        Carnegie Mellon University   United States   
120   121                   Queen Mary University of London  United Kingdom   
39     40                           University of Hong Kong       Hong Kong   

       Region  Total_faculty International_faculty  Total_students  \
24     Europe            825                  None           10065   
178    Europe            340                  None            4969   
7      Europe           1391                  None           15857   
37     Europe            886                  None            9928   
102    Europe            929                  None           16727   
15     Europe           2886                  None           30304   
143    Europe            693                  None            9005   
23   Americas            939                  None           12676   
120    Europe           1158                  None           15405   
39       Asia           1020                  None           18364   

     International_students  Internationality_ratio  
24                     7146                0.709985  
178                    2832                0.569934  
7                      8721                0.549978  
37                     5460                0.549960  
102                    8364                0.500030  
15                    14849                0.490001  
143                    4322                0.479956  
23                     5704                0.449984  
120                    6932                0.449984  
39                     7713                0.420007

Both sources confirm that *London School of Economics and Political Science* is at the top of being the most international, with a large 70% of students being from outside of the UK. Among the top 10 on both lists are also:
* EPFL
* Carnegie Mellon University
* Imperial College London
* University Colege London
* Maastricht University
* Queen Mary University of London
* University of St. Andrews
* University of Hong Kong

University of Luxembourg is missing in the second list event though it is very highly ranked in the first one according to the above mentioned criterion.

** Best universities in terms of: ratio between faculty members and students grouped by country **

In [81]:
# Group the data according to the value of the country attribute
# Calculate the ratio between the number of students and staff members, sort the dataframe according to that value
find_ratio(times_ranking_df, 'Total_students', 'Total_faculty', 'Student_teacher_ratio', 'Country', ascending=True).head(5)

Total_students  Total_faculty  Student_teacher_ratio
Country                                                                 
Denmark                      64110          10598               6.049255
Russian Federation           29236           4005               7.299875
Japan                        48481           6465               7.498995
Italy                         1205            158               7.626582
Switzerland                 107852          10049              10.732610

Even though there is a big difference between the university faculty numbers from both data sources, both lists agree that the top 2 countries according to this criterion are Denmark and Russia.

** Best universities in terms of: ratio of international students grouped by country **

In [85]:
# Group the data according to the value of the country attribute
# Calculate the ratio between the number of international students and total number of students
# sort the dataframe according to that value
find_ratio(times_ranking_df, 'International_students', 'Total_students', 'Internationality_ratio', 'Country').head(5)

International_students  Total_students  Internationality_ratio
Country                                                                       
Luxembourg                        2832            4969                0.569934
United Kingdom                  213058          596449                0.357211
Hong Kong                        25158           77663                0.323938
Australia                        83812          268630                0.311998
Singapore                        17086           56101                0.304558

The United Kingdom and Hong Kong are second and third on both lists with a very similar ratio, respectively. Luxembourg has a huge internationality ratio according to the first source, but it didn't make it to the list in the second. Australia is also among the top 5 on both rankings, but the ratio differs by a few points.

** Best universities in terms of: ratio between faculty members and students grouped by region **

In [89]:
# Group the data according to the value of the region attribute
# Calculate the ratio between the number of students and staff members, sort the dataframe according to that value
find_ratio(times_ranking_df, 'Total_students', 'Total_faculty', 'Student_teacher_ratio', 'Region', ascending=True)

Total_students  Total_faculty  Student_teacher_ratio
Region                                                        
Africa             20775           1776              11.697635
Americas         1954934         165594              11.805585
Asia              530949          42636              12.453068
Europe           2143623         126864              16.897016
Oceania           298978          11554              25.876580

Note that there are two reasons this table is difficult to comment:
1. the large difference in the raw data about staff members for each university, which makes figures difficult to compare
2. the second source aggregates North and South America into a single region, so the values are aggregated

** Best universities in terms of: ratio of international students grouped by region **

In [90]:
# Group the data according to the value of the region attribute
# Calculate the ratio between the number of international students and total number of students, sort the dataframe according to that value
find_ratio(times_ranking_df, 'International_students', 'Total_students', 'Internationality_ratio', 'Region')

International_students  Total_students  Internationality_ratio
Region                                                                  
Oceania                    92613          298978                0.309765
Europe                    481576         2143623                0.224655
Africa                      3740           20775                0.180024
Americas                  336271         1954934                0.172011
Asia                       78839          530949                0.148487

The aggregated tables according to the ratio of international students are pretty much identical from both sources, with some small differences in the ratio values. Once North America and South America are averaged in the first case, it will easily be seen a ratio close to 0.17 is obtained.

Overall, the most noticeable characteristic in the data is the big discrepancy between the student-to-faculty ratio computed from the first source and the second one. And since there is no staff number information in the second source, it's impossible to detect the source or reason for this big difference.

In the second dataset, there is a tendency of a very large student-to-staff ratio value, which after doing the calculations results in significantly fewer staff members per university, as compared to the first source. Therefore, it is understandable that all the derived and aggregated results will differ, somewhere even relatively greatly. However, there is no way to judge which of the two is more reliable without obtaining additional information from another external source.